### PR TITLE
Converting HttpServerBuilder to an interface

### PR DESCRIPTION
**Pull request by chemicL**
Motivation:

Following the changes in `HttpClientBuilder`s with the end goal of
achieving the API design consistency, the `HttpServerBuilder` has been
converted to an interface. All the intermediate base abstract class
implementation has been pushed down to the `http-netty` package.

Modifications:

- `HttpServerBuilder` converted to interface,
- `OffloadingFilter` package private class moved to `http-netty` package,
- `ClearAsyncContextHttpServiceFilter` and
  `StrategyInfluencerAwareConversions` were made public in the
  `http-api` package as they provide useful functionality to those
  willing to implement their own `HttpServerBuilder`.

Result:

Consistent API with interfaces instead of implementation scattered
between final class and the base abstract class.

### CODE REVIEW

File: servicetalk-http-api/src/main/java/io/servicetalk/http/api/ClearAsyncContextHttpServiceFilter.java
Comment (idelpivnitskiy) [inline servicetalk-http-api/src/main/java/io/servicetalk/http/api/ClearAsyncContextHttpServiceFilter.java:24]: Discussed offline: this class doesn't have to be public, it can move to the `http-netty` module as pkg-private

File: servicetalk-http-api/src/main/java/io/servicetalk/http/api/HttpServerBuilder.java
Comment (idelpivnitskiy) [inline servicetalk-http-api/src/main/java/io/servicetalk/http/api/HttpServerBuilder.java:153]: Shifted indentation

File: servicetalk-http-api/src/main/java/io/servicetalk/http/api/HttpServerBuilder.java
Comment (idelpivnitskiy) [inline servicetalk-http-api/src/main/java/io/servicetalk/http/api/HttpServerBuilder.java:158]: Please, don't mix this PR with unrelated changes. Consider submitting a separate PR to clean up deprecations. Similar to what Scott did: https://github.com/apple/servicetalk/pull/1805
It will help to highlight all meaningful changes as separate items on release notes, and keep the scope of each PR independent.

File: servicetalk-http-api/src/main/java/io/servicetalk/http/api/HttpServerBuilder.java
Comment (idelpivnitskiy) [inline servicetalk-http-api/src/main/java/io/servicetalk/http/api/HttpServerBuilder.java:268]: Shifted indentation

File: servicetalk-http-api/src/main/java/io/servicetalk/http/api/HttpServerBuilder.java
Comment (idelpivnitskiy) [inline servicetalk-http-api/src/main/java/io/servicetalk/http/api/HttpServerBuilder.java:312]: Shifted indentation

File: servicetalk-http-api/src/main/java/io/servicetalk/http/api/HttpServerBuilder.java
Comment (idelpivnitskiy) [inline servicetalk-http-api/src/main/java/io/servicetalk/http/api/HttpServerBuilder.java:329]: As described in #1819, we don't need to add default implementations for these methods back

File: servicetalk-http-api/src/main/java/io/servicetalk/http/api/StrategyInfluencerAwareConversions.java
Comment (idelpivnitskiy) [inline servicetalk-http-api/src/main/java/io/servicetalk/http/api/StrategyInfluencerAwareConversions.java:28]: Summary from offline discussion: we don't need to make this utility class public. Methods that are required for the server builder impl can be moved to `http-netty` as pkg-private.

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultHttpServerBuilder.java
Comment (idelpivnitskiy) [inline servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultHttpServerBuilder.java:105]: Shifted indentation

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultHttpServerBuilder.java
Comment (idelpivnitskiy) [inline servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultHttpServerBuilder.java:121]: Shifted indentation

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultHttpServerBuilder.java
Comment (idelpivnitskiy) [inline servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultHttpServerBuilder.java:232]: All methods that were `protected` can be `private` now

File: servicetalk-http-api/src/main/java/io/servicetalk/http/api/HttpServerBuilder.java
Comment (idelpivnitskiy) [inline servicetalk-http-api/src/main/java/io/servicetalk/http/api/HttpServerBuilder.java:504]: For any public methods that you remove from this API, consider submitting a PR for 0.41 branch to mark those `@Deprecated`. For cases when we do not provide an alternative API (like here), consider clarifying that users have to do it themselves. Example: #1791.

File: servicetalk-http-api/src/main/java/io/servicetalk/http/api/HttpServerBuilder.java
Comment (idelpivnitskiy) [inline servicetalk-http-api/src/main/java/io/servicetalk/http/api/HttpServerBuilder.java:48]: Consider moving all these `static` utilities to `http-netty` as private or pkg-private utils. Even if users implement this builder interface, it's questionable if they need these utils. We can always make them public later, if there is a good use-case for that.

**Review by idelpivnitskiy — CHANGES_REQUESTED** [review]
Discussed offline and agreed to defer merging this PR until after #1824 is merged to allow cherry-picking observability API into 0.41 without conflicts.

File: servicetalk-http-api/src/main/java/io/servicetalk/http/api/ClearAsyncContextHttpServiceFilter.java
Comment (bondolo) [reply servicetalk-http-api/src/main/java/io/servicetalk/http/api/ClearAsyncContextHttpServiceFilter.java:24]: We should at some point consider an `http-internal` module for the non-Netty specific HTTP internals.

**Review by bondolo — APPROVED** [review]
Once the issues @idelpivnitskiy mentions are resolved, LGTM!

File: servicetalk-http-api/src/main/java/io/servicetalk/http/api/ClearAsyncContextHttpServiceFilter.java
Comment (idelpivnitskiy) [reply servicetalk-http-api/src/main/java/io/servicetalk/http/api/ClearAsyncContextHttpServiceFilter.java:24]: The new module doesn't make much sense if classes from that module are not used by other 2+ modules. For `http-netty` only, it's more restrictive to use pkg-private classes.
There are a few other utilities that will be nice to share between `http-api` and `http-netty`, but there is "chicken and egg" problem as those utilities depend on `http-api` 😕

File: servicetalk-http-api/src/main/java/io/servicetalk/http/api/HttpServerBuilder.java
Comment (chemicL) [reply servicetalk-http-api/src/main/java/io/servicetalk/http/api/HttpServerBuilder.java:504]: Please check https://github.com/apple/servicetalk/pull/1830

File: servicetalk-http-api/src/main/java/io/servicetalk/http/api/HttpServerBuilder.java
Comment (chemicL) [reply servicetalk-http-api/src/main/java/io/servicetalk/http/api/HttpServerBuilder.java:158]: Brought back the method. Removal can follow after this is merged.

Comment (idelpivnitskiy) [reply servicetalk-http-api/src/main/java/io/servicetalk/http/api/HttpServerBuilder.java:158]: I see that you have the same unrelated change in client builder PR: https://github.com/apple/servicetalk/pull/1819/files#diff-f220503d52fc93c626fb3cbfc3a3814f449420c90bc980ae509d40af7e7de085L100
Consider cleaning up all deprecated methods (for client/server http/grpc) builders before these conversions. Going forward, we should clean up deprecated methods as soon as those are cherry-picked into 0.41.

Comment (chemicL) [reply servicetalk-http-api/src/main/java/io/servicetalk/http/api/HttpServerBuilder.java:158]: Agreed, please approve: https://github.com/apple/servicetalk/pull/1844

**Review by idelpivnitskiy — APPROVED** [review]
LGTM after #1844, assuming merge conflicts don't lead to major changes. Thanks for waiting for observability PR!

File: servicetalk-http-api/src/main/java/io/servicetalk/http/api/HttpServerBuilder.java
Comment (idelpivnitskiy) [inline servicetalk-http-api/src/main/java/io/servicetalk/http/api/HttpServerBuilder.java:271]: On the `SingleAddressHttpClientBuilder` interface a similar method has `default` implementation. Consider making client and server API consistent and either use `default` in both cases or let the implementation handle it. I don't have a preference.

Comment (chemicL) [reply servicetalk-http-api/src/main/java/io/servicetalk/http/api/HttpServerBuilder.java:271]: The PR that addresses this suggestion: https://github.com/apple/servicetalk/pull/1866

**Pull request merged**
